GH-3838 Add Support for S3DownloadStrategy to Bottles

eedwards-sk opened the pull request:
Add Support for S3BottleDownloadStrategy to Bottles

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew style` with your changes locally?

```
→ brew style
warning: parser/current is loading parser/ruby23, which recognizes
warning: 2.3.6-compliant syntax, but you are running 2.3.3.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
Inspecting 665 files
.........................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................

665 files inspected, no offenses detected
```

- [x] Have you successfully run `brew tests` with your changes locally?

```
2246 examples, 0 failures, 10 pendings

Took 104 seconds (1:44)
```

-----

#### Preface

No tests written yet. I am happy to write them if we determine this approach is acceptable.
I am taking your contribution notes at heart, and wish to deliver an improvement that fits within the guidelines of the quality bar Homebrew has set. Thus, I expect to iterate on this, as needed, until hopefully we can determine what all changes are needed.

#### Overview

This is an attempt at implementing #3838 by allowing for a `:using =>` statement to be optionally added to the `root_url` of a bottle configuration.

I can specify a new download strategy called `S3BottleDownloadStrategy` which extends the behavior of the `CurlBottleDownloadStrategy` by allowing it to use S3 URLs, but still retaining the bottle "Pouring" text.

#### Example

```
class MyApp < Formula
  desc "MyOrg's MyApp"
  homepage "https://github.com/myorg/myapp"
  url "https://myorg-private-source-dist.s3.amazonaws.com/myapp/1.0/my-app-1.0.0.tar.gz", :using => S3DownloadStrategy
  version "1.0.0"
  sha256 "big-long-sha"
  
  bottle do
    root_url "https://myorg-private-bin-repo.s3.amazonaws.com/myapp/1.0", :using => S3BottleDownloadStrategy
    cellar :any_skip_relocation
    sha256 "big-long-sha" => :high_sierra
  end
  
  def install
    system "./configure", "--disable-silent-rules",
                          "--prefix=#{prefix}"
    system "make", "install"
  end
  
  test do    
    true
  end
end
```

#### Code Changes

##### Library/Homebrew/download_strategy.rb

`CurlBottleDownloadStrategy` is the default bottle download strategy. To support S3, I needed a new class which inherits from this one, and adds the S3 Behavior.

- Added `S3BottleDownloadStrategy` which inherits from `CurlBottleDownloadStrategy`
- Moved the `_fetch` override from `S3DownloadStrategy` to a new mixin inside `CurlDownloadStrategy` called `S3DownloadExtension`
- added `include S3DownloadExtension` into `S3DownloadStrategy` and `S3BottleDownloadStrategy`

##### Library/Homebrew/software_spec.rb

`BottleSpecification` has a `root_url` attribute, I needed to be able to pass a `:using =>` setting into it, so I used a pattern similar to `resource.url` to support optionally passing in a hash of 'specs'.

I then store those specs in a new read-only attribute called `:root_url_specs`

```
-  def root_url(var = nil)
+  def root_url(var = nil, specs = {})
     if var.nil?
       @root_url ||= "#{DEFAULT_DOMAIN}/#{Utils::Bottles::Bintray.repository(tap)}"
     else
       @root_url = var
+      @root_url_specs.merge!(specs)
     end
   end
```

`Bottle.initialize` normally defaults to `CurlBottleDownloadStrategy` and ignores whatever `resource.url` had set it to. I needed to be able to remove the default override and instead use a default suggestion to `resource.url` which would allow the new `S3BottleDownloadStrategy` to be optionally specified.

```
-    @resource.url(build_url(spec.root_url, filename))
-    @resource.download_strategy = CurlBottleDownloadStrategy
```
```
+    @resource.url(build_url(spec.root_url, filename),
+                  select_download_strategy(spec.root_url_specs))
```

I added a helper method `select_download_strategy` which takes the `root_url_specs` hash from the `BottleSpecification`

It's here that I default to `CurlBottleDownloadStrategy`, except now I'm doing it by specifying it as a `:using` flag, and this allows the existing `Formula.url` behavior to set `@resource.download_strategy` appropriately.

```
+  def select_download_strategy(specs)
+    specs[:using] = CurlBottleDownloadStrategy unless specs.include?(:using)
+    specs
+  end
```

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -450,6 +483,18 @@ def stage
+# to sign the request. This strategy is good in a corporate setting,
+# because it lets you use a private S3 bucket as a repo for
+# internal distribution. (It will work for public buckets as well.)
+class S3BottleDownloadStrategy < CurlBottleDownloadStrategy
```
It would be good if there was a way the bottle download strategy code could be refactored such that this could use any existing `:using` without needing a code change for each. For example, bottles could be handled by `CurlDownloadStrategy` by default with a `bottle: true` argument or something.

eedwards-sk replied:
Thanks Mike, I'll look into this. Not sure how to do it yet cause I'm anything but a ruby wizard, but I'll see if I can get something to look at :)

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -397,6 +397,39 @@ def curl(*args, **options)
+      begin
+        require "aws-sdk-v1"
+      rescue LoadError
+        onoe "Install the aws-sdk gem into the gem repo used by brew."
```
Could use `def install_gem_setup_path!` for this (see `utils.rb`)

eedwards-sk replied:
Thanks for the comment. This `_fetch` behavior is completely unchanged (I literally just cut and paste it into the new module).

I'm happy to look into improving it though, thanks!

eedwards-sk commented:
Hey @MikeMcQuaid,

Thanks again for your comments.

I've attempted to address them. Let me describe the changes for your review:

#### Code Changes

##### Library/Homebrew/download_strategy.rb

- `class AbstractDownloadStrategy`

I noticed that the `resource.specs` were getting saved to the download strategy as `meta`, so I'm leveraging that to store a new `:bottle` flag.

During initialize, `AbstractDownloadStrategy` will `extend BottleExtension` if `meta[:bottle] == true` (This adds the `stage` behavior that was previously in `CurlBottleDownloadStrategy` and `LocalBottleDownloadStrategy`)

Thus, in other contexts when the download strategy is being initialized, the caller can specify `resource.specs[:bottle] = true` before passing the resource to that strategy's constructor.

- `class CurlDownloadStrategy`

Removed my `S3DownloadExtension` module, as it was moved back into the original `S3DownloadStrategy` class.

- `class CurlBottleDownloadStrategy`

Removed this class entirely since it's now covered by `CurlDownloadStrategy` with `meta[:bottle] = true`

- `class LocalBottleDownloadStrategy`

This class is still used to allow the cached path to be specified during creation, so I left this as-is other than removing the `stage` behavior which was duplicated across both this and `CurlBottleDownloadStrategy` (it's now covered in `AbstractDownloadStrategy` with `extend BottleExtension`)

- `class S3DownloadStrategy`

Moved the `_fetch` behavior back into this class, since this strategy is now valid for both regular formula `url`s as well as a bottle `root_url`

Also as per your comments, created a helper function `load_aws_sdk` which leverages `Homebrew.install_gem_setup_path!` to automatically install the `aws-sdk-v1` gem.

##### Library/Homebrew/constants.rb

To match the other uses of `Homebrew.install_gem_setup_path!`, I've added a new constant to specify the gem version for the `aws-sdk-v1` gem (used pessimistic matching to match any version within 1.x)

##### Library/Homebrew/software_spec.rb

- `class Bottle`

I now specify `@resource.specs[:bottle] = true` when creating the resource, so that the `DownloadStrategy` will end up extending with `BottleExtension`

Updated the private helper `select_download_strategy` so that it will default to `CurlDownloadStrategy`, since this strategy now works for bottles and non-bottles (depending on the meta bottle flag).

##### Library/Homebrew/formulary.rb

- `class BottleLoader`

**This is the most confusing part**, because this code wasn't getting exercised when I was making the change, but I found the reference to `CurlBottleDownloadStrategy` here, so I updated it to instead use `CurlDownloadStrategy` with a resource that has `specs[:bottle] = true`.

**What I'm not sure of is the context in which this gets called??**

#### New Example

Updated example usage, notice that I can now use `S3DownloadStrategy` for both the url and for bottle `root_url`:

```
class MyApp < Formula
  desc "MyOrg's MyApp"
  homepage "https://github.com/myorg/myapp"
  url "https://myorg-private-source-dist.s3.amazonaws.com/myapp/1.0/my-app-1.0.0.tar.gz", :using => S3DownloadStrategy
  version "1.0.0"
  sha256 "big-long-sha"
  
  bottle do
    root_url "https://myorg-private-bin-repo.s3.amazonaws.com/myapp/1.0", :using => S3DownloadStrategy
    cellar :any_skip_relocation
    sha256 "big-long-sha" => :high_sierra
  end
  
  def install
    system "./configure", "--disable-silent-rules",
                          "--prefix=#{prefix}"
    system "make", "install"
  end
  
  test do    
    true
  end
end
```

MikeMcQuaid commented:
Great work @eedwards-sk, I really like your approach and glad I encouraged this as it's a massive improvement. Would love to see a few tests to cover this behaviour.

eedwards-sk commented:
Thanks for the feedback. I'm currently working on adding tests. Unfortunately, there weren't any tests for S3DownloadStrategy before, at all, so as I'm attempting to add coverage I'm starting to question the sanity of the current strategy. It feels a bit difficult to test...

We're basically putting a contract on `S3DownloadStrategy#fetch` that if the aws sdk gem is not installed, it will get installed.

However, testing this is extremely difficult. I don't want to mock a bunch of the internals with fragile tests, I just want to validate the contract.

Ideally I'd be able to set up a clean, repeatable state e.g. given an environment without the aws sdk gem, confirm that after calling `subject._fetch`, it gets installed. (except it would get installed once and the state would be dirty...)

Since so much of the 'gem environment' is dynamic and set up by invoking through `brew tests`, I can't really 'mock' the environment.

To make things even more confusing, the dynamic install and immediate require works when running directly through a `brew install`, but when running through rspec (well, `brew tests`) it fails to require the gem after installation:

code
```
def _fetch
...
    require "rubygems"
    aws_sdk_s3 = "aws-sdk-s3"
    Homebrew.install_gem_setup_path! aws_sdk_s3, HOMEBREW_AWS_SDK_S3_VERSION, nil
    require aws_sdk_s3
```

rspec
```
     Failure/Error: subject._fetch
     
     LoadError:
       cannot load such file -- aws-sdk-s3
```

Finally, to properly create aws sdk stubbed responses, I'll need the aws sdk gem (now the `aws-sdk-s3` gem) in the rspec environment, which means adding it to the `test/Gemfile`... doing this also causes the gem to already be installed into the vendored ruby's gems, again ruining the ability to test if the dynamic install will work.

I guess it would also help if there was already coverage for `Homebrew.install_gem_setup_path!`, as I really don't want to test whether or not it actually works, I just want to ensure that our `_fetch` method will call it.

Basically, trying to figure out if I need to extract something here into a separate util method or not. The struggle with that is doing it and making it testable while maintaining the 'dynamic gem installation' behavior.

I'll keep hacking away at it, but if you have any thoughts please feel free to share.

eedwards-sk commented:
The gem environments are definitely changing in a way I'm not clear on.

If I add a line with
```
raise RuntimeError, `gem list`
```
Immediately after the call to `Homebrew.install_gem_setup_path!`

and then I run a `brew install`...

```
==> Installing or updating 'aws-sdk-s3' gem
Fetching: jmespath-1.3.1.gem (100%)
Successfully installed jmespath-1.3.1
Fetching: aws-partitions-1.68.0.gem (100%)
Successfully installed aws-partitions-1.68.0
Fetching: aws-sigv4-1.0.2.gem (100%)
Successfully installed aws-sigv4-1.0.2
Fetching: aws-sdk-core-3.17.0.gem (100%)
Successfully installed aws-sdk-core-3.17.0
Fetching: aws-sdk-kms-1.5.0.gem (100%)
Successfully installed aws-sdk-kms-1.5.0
Fetching: aws-sdk-s3-1.8.2.gem (100%)
Successfully installed aws-sdk-s3-1.8.2
6 gems installed
Error: aws-partitions (1.68.0)
aws-sdk-core (3.17.0)
aws-sdk-kms (1.5.0)
aws-sdk-s3 (1.8.2)
aws-sigv4 (1.0.2)
bigdecimal (1.2.8)
bundler (1.16.1)
byebug (10.0.0)
CFPropertyList (2.2.8)
coderay (1.1.2)
did_you_mean (1.0.0)
io-console (0.4.5)
jmespath (1.3.1)
json (1.8.3)
libxml-ruby (2.9.0)
method_source (0.9.0)
minitest (5.8.5)
net-telnet (0.1.1)
nokogiri (1.5.6)
power_assert (0.2.6)
pry (0.11.3)
pry-byebug (3.6.0)
psych (2.1.0)
rake (10.4.2)
rdoc (4.2.1)
sqlite3 (1.3.11)
test-unit (3.1.5)
```

but shortly thereafter, as it thinks that `Bottle installation failed: building from source.`, it ends up using `S3DownloadStrategy` for the source, as well, which when it dumps `gem list` results in:

```
Error: aws-partitions (1.68.0)
aws-sdk-core (3.17.0)
aws-sdk-kms (1.5.0)
aws-sdk-s3 (1.8.2)
aws-sigv4 (1.0.2)
bigdecimal (1.2.8)
io-console (0.4.5)
jmespath (1.3.1)
json (1.8.3)
psych (2.1.0)
rdoc (4.2.1)
```

Much different already, even in the same run!

Then if I run it from `brew tests`:

```
     Failure/Error: subject._fetch
     
     RuntimeError:
       ast (2.3.0)
       bundler (1.16.1)
       codecov (0.1.10)
       diff-lcs (1.3)
       docile (1.1.5)
       json (2.1.0)
       parallel (1.12.0)
       parallel_tests (2.17.0)
       parser (2.4.0.2)
       powerpack (0.1.1)
       rainbow (3.0.0)
       rspec (3.6.0)
       rspec-core (3.6.0)
       rspec-expectations (3.6.0)
       rspec-its (1.2.0)
       rspec-mocks (3.6.0)
       rspec-retry (0.5.6)
       rspec-support (3.6.0)
       rspec-wait (0.0.9)
       rubocop (0.52.1)
       ruby-progressbar (1.9.0)
       simplecov (0.15.1)
       simplecov-html (0.10.2)
       unicode-display_width (1.3.0)
       url (0.3.2)
```

Which is totally different. So I thought, hmm... maybe a difference in environment? Well, let's switch the bail to `gem environment`...

```
RubyGems Environment:
  - RUBYGEMS VERSION: 2.5.2
  - RUBY VERSION: 2.3.3 (2016-11-21 patchlevel 222) [universal.x86_64-darwin17]
  - INSTALLATION DIRECTORY: /usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.3.0
  - USER INSTALLATION DIRECTORY: /Users/eedwards/.gem/ruby/2.3.0
  - RUBY EXECUTABLE: /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin/ruby
  - EXECUTABLE DIRECTORY: /usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.3.0/bin
  - SPEC CACHE DIRECTORY: /Users/eedwards/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /Library/Ruby/Site
  - RUBYGEMS PLATFORMS:
    - ruby
    - universal-darwin-17
  - GEM PATHS:
     - /usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.3.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /usr/local/Homebrew/Library/Homebrew/vendor/bundle/ruby/2.3.0/bin
     - /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin
     - /usr/bin
     - /bin
     - /usr/sbin
     - /sbin
     - /usr/local/Homebrew/Library/Homebrew/shims/scm
```

It's identical in all 3 cases, except for `     - /usr/local/Homebrew/Library/Homebrew/shims/`... which is `scm` for the first output during `brew tests`, which then changes to `super` for the second output. Through rspec, it's `scm` again. That's the only difference, is that related or a red herring?

eedwards-sk commented:
I've tracked down part of the issue, when running `brew tests`, you're inside a 'bundle', and the newly installed gems inside the 'homebrew setup path' aren't seen, as they are rejected for not being a part of the 'bundle'.

The environments, otherwise, appear about the same.

I'm going to investigate if there's a way that I can get the bundled environment that rspec runs within to recognize the new gem after it's installed.

Once I can validate that, then I can probably start to extract away some of the 'install gem setup path' behavior into separate methods.

Then I can implement testing around 'dynamically install a gem into ANY gem repo' (thus setup and validate against an isolated test repo), and just make sure that 'install gem setup path' calls those functions with the correct args.

eedwards-sk commented:
I've now confirmed that when building from source, due to the way that the environment is sanitized, that we cannot easily automatically install a gem (it fails).

I will do some minor digging into seeing if there's a simple way to make it work, otherwise I will likely back out from attempting to use `install_gem_setup_path!` which is seeming to be not the ideal solution for this use case (and I will go back to the prior behavior that was implemented, which requires the gem to already be installed).

To reiterate, the following issues crop up due to attempting to dynamically install a gem:

- The bottle fetching environment is different from the build environment
  - installing the gem inside the bottling phase means it will work in the build phase, BUT...
  - installing the gem in the build phase does not work
- rspec runs within a bundle, meaning the gem list is limited to those in the Gemfile it's using
  - running with a `bundle.clean_env` doesn't fix it, since bundle already overwrote a bunch of Gem commands and expects a Gemfile
  - running a `bundle install` wouldn't work, since it needs to occur after the gem is installed, but before it's required... it also wouldn't propagate to the `build` run

eedwards-sk commented:
@MikeMcQuaid It unfortunately does appear that `Homebrew.install_gem_setup_path!` cannot be a solution for this use case, as it, during the build phase, runs into Mac System Integrity Protection (SIP) issues, which appears to be similar to #3849

```
==> Installing or updating 'aws-sdk-s3' gem
Error: Operation not permitted @ dir_s_mkdir - /Users/eedwards/.gem
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/fileutils.rb:253:in `mkdir'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/fileutils.rb:253:in `fu_mkdir'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/fileutils.rb:227:in `block (2 levels) in mkdir_p'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/fileutils.rb:225:in `reverse_each'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/fileutils.rb:225:in `block in mkdir_p'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/fileutils.rb:211:in `each'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/fileutils.rb:211:in `mkdir_p'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/source.rb:157:in `fetch_spec'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/resolver/api_specification.rb:77:in `spec'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/resolver/installer_set.rb:78:in `add_always_install'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/dependency_installer.rb:474:in `resolve_dependencies'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/commands/install_command.rb:239:in `install_gem'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/commands/install_command.rb:300:in `block in install_gems'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/commands/install_command.rb:296:in `each'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/commands/install_command.rb:296:in `install_gems'
/System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/lib/ruby/2.3.0/rubygems/commands/install_command.rb:203:in `execute'
/usr/local/Homebrew/Library/Homebrew/utils.rb:227:in `install_gem_setup_path!'
/usr/local/Homebrew/Library/Homebrew/download_strategy.rb:507:in `install_aws_s3_sdk'
/usr/local/Homebrew/Library/Homebrew/download_strategy.rb:476:in `initialize'
/usr/local/Homebrew/Library/Homebrew/resource.rb:59:in `new'
/usr/local/Homebrew/Library/Homebrew/resource.rb:59:in `downloader'
/usr/local/Homebrew/Library/Homebrew/resource.rb:137:in `fetch'
/usr/local/Homebrew/Library/Homebrew/resource.rb:90:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1872:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1090:in `brew'
/usr/local/Homebrew/Library/Homebrew/build.rb:111:in `block in install'
/usr/local/Homebrew/Library/Homebrew/utils.rb:552:in `with_env'
/usr/local/Homebrew/Library/Homebrew/build.rb:108:in `install'
/usr/local/Homebrew/Library/Homebrew/build.rb:189:in `<main>'
```

from ps during a run:
```
73889 ttys003    0:00.04 /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/build.rb /usr/local/Homebrew/Library/Taps/myorg/homebrew-private/Formula/my-app-1.0.rb --debug
```

alyssais commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -502,6 +492,22 @@ def _fetch
+      require aws_sdk_gem
+    rescue LoadError
+      raise if retries.positive?
+      Homebrew.install_gem_setup_path! aws_sdk_gem, HOMEBREW_AWS_SDK_V1_VERSION, aws_sdk_gem_exe
```
You should be able to just always use `install_gem_setup_path!`
 — that’s how we do it everywhere else. Would make this method much simpler.

eedwards-sk replied:
@alyssais Thanks! I did switch to that approach before I discovered I had to abandon the use of the method entirely :(

eedwards-sk commented:
So the issues I'm having with using `install_gem_setup_path!` in the `build` phase don't seem to be SIP related after all, but are likely due to me now realizing that it's using `/usr/bin/sandbox-exec` when it runs, which likely is stopping it from being able to write to the user directory.

eedwards-sk commented:
@MikeMcQuaid

Please see latest changes :)

I have refactored `S3DownloadStrategy` to use the latest AWS SDK Gem (v3), and more specifically, to use their new 'service level' gem, which allows you to require only the gem for the specific service you're using. Thus in this case, it now requires `aws-sdk-s3`.

In addition, this gem has been added to the test `Gemfile`, to satisfy the rspec runs, but it is still not a requirement for regular brew use. (Added the v1 gem to the Gemfile would have lead to a conflict on the current `json` version and a bunch more gem version changes, so using the newer gem actually has a much smaller impact!)

As there was no test coverage for this strategy before, I've now added unit test coverage. To do so I had to split apart the behavior a bit. It probably could use another refactor pass later at some point, or even moving some of the AWS / S3 related behavior to dedicated classes with their own testing. However, this should at least ensure the initial contract cases are covered.

TODO: Add integration testing for the Bottle / not bottle behavior in `AbstractDownloadStrategy`

eedwards-sk commented:
Added unit tests for the bottle / pourable behavior in AbstractDownloadStrategy.

Please review for a possible merge.

Let me know if you have any questions, comments, concerns, etc. :)

```
→ brew style
warning: parser/current is loading parser/ruby23, which recognizes
warning: 2.3.6-compliant syntax, but you are running 2.3.3.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
Inspecting 668 files
<snip>

668 files inspected, no offenses detected
```
```
→ brew tests
Randomized with seed 33573
8 processes for 223 specs, ~ 27 specs per process
warning: parser/current is loading parser/ruby23, which recognizes
warning: 2.3.6-compliant syntax, but you are running 2.3.3.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
warning: parser/current is loading parser/ruby23, which recognizes
warning: 2.3.6-compliant syntax, but you are running 2.3.3.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
warning: parser/current is loading parser/ruby23, which recognizes
warning: 2.3.6-compliant syntax, but you are running 2.3.3.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
warning: parser/current is loading parser/ruby23, which recognizes
warning: 2.3.6-compliant syntax, but you are running 2.3.3.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
warning: parser/current is loading parser/ruby23, which recognizes
warning: 2.3.6-compliant syntax, but you are running 2.3.3.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
warning: parser/current is loading parser/ruby23, which recognizes
warning: 2.3.6-compliant syntax, but you are running 2.3.3.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.
warning: parser/current is loading parser/ruby23, which recognizes
warning: 2.3.6-compliant syntax, but you are running 2.3.3.
warning: please see https://github.com/whitequark/parser#compatibility-with-ruby-mri.

Finished in 17.49 seconds (files took 2.04 seconds to load)
452 examples, 0 failures, 3 pending

Finished in 31.49 seconds (files took 1.89 seconds to load)
232 examples, 0 failures, 1 pending

Finished in 31.9 seconds (files took 1.92 seconds to load)
278 examples, 0 failures, 1 pending

Finished in 59.18 seconds (files took 1.9 seconds to load)
234 examples, 0 failures

Finished in 1 minute 0.33 seconds (files took 1.89 seconds to load)
259 examples, 0 failures

Finished in 1 minute 0.8 seconds (files took 1.8 seconds to load)
289 examples, 0 failures

Finished in 1 minute 14.52 seconds (files took 2 seconds to load)
301 examples, 0 failures, 2 pending

Finished in 1 minute 46.43 seconds (files took 1.88 seconds to load)
229 examples, 0 failures, 3 pending

2274 examples, 0 failures, 10 pendings

Took 108 seconds (1:48)
```

eedwards-sk commented:
I noticed that build was failing, discovered a bug in the original implementation of `S3DownloadStrategy` that was not surfaced until I added tests.

The behavior that was attempting to use `object_summary.public_url` would never have been exercised in the real world.

Credentials are actually required to create an s3 object summary, thus you wouldn't be able to instantiate an object summary and call `public_url` on it.

Thus I've defaulted the behavior in that case to instead just use the `url` as-is, as it's the same url you would get back from the 'authenticated' `public_url` method, anyway.

Theoretically the only use case for this behavior would be specifying a download with `:using => S3DownloadStrategy`, but the user doesn't have any aws credentials configured, and the resulting S3 url ends up being public anyway so it didn't need them.

That is a rare use case though, since if it's a public S3 url you want to just use the normal CURL downloader, instead (and thus not use `:using => S3DownloadStrategy`).

MikeMcQuaid commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -15,6 +15,7 @@ def initialize(name, resource)
     @url = resource.url
     @version = resource.version
     @meta = resource.specs
+    extend Pourable if meta.key?(:bottle) && meta[:bottle]
```
When would `meta[:bottle]` be truthy but not `meta.key?(:bottle)` (I think never, no?)

eedwards-sk replied:
Ah thanks, I switch languages a lot, I thought I was protecting against an exception on accessing a non-existent key.

MikeMcQuaid replied:
No worries, we've all been there!

MikeMcQuaid commented on Library/Homebrew/test/Gemfile:
```diff
@@ -2,6 +2,7 @@ source "https://rubygems.org"
 
 require_relative "../constants"
 
+gem "aws-sdk-s3"
```
Can the relevant objects be stubbed/mocked rather than needing to add the gem at test time?

eedwards-sk replied:
Yes, instead of turning on stubs for the S3 library I could instead mock the S3 library interface itself, which would no longer exercise any of the S3 library itself during testing.

MikeMcQuaid replied:
Yeh, that sounds like a good call for this case, thanks!

MikeMcQuaid commented:
> So the issues I'm having with using install_gem_setup_path! in the build phase don't seem to be SIP related after all, but are likely due to me now realizing that it's using /usr/bin/sandbox-exec when it runs, which likely is stopping it from being able to write to the user home directory.

How about trying to install it in advance e.g. in `formula_installer.rb`?

eedwards-sk commented on Library/Homebrew/download_strategy.rb:
```diff
@@ -472,35 +467,66 @@ def stage
+    require "aws-sdk-s3"
+  rescue LoadError
+    onoe "please install the aws-sdk-s3 gem into the gem repo used by brew"
+    onoe "example: gem install aws-sdk-s3 --no-rdoc --no-ri --user-install"
```
@amancevice sorry, can you please clarify? You're suggesting to remove line 502?

amancevice replied:
No no, just asking a question. I have been banging my head against the wall with this problem as well -- But it turns out I misunderstood what the `--user-install` flag was doing.

I was originally confused by the error message `please install the aws-sdk-s3 gem into the gem repo used by brew` because I had no idea where to find this repo.

Please disregard.

amancevice replied:
That said, after running `gem install aws-sdk-s3 --no-rdoc --no-ri --user-install`, I still get a `LoadError` when I try to require the `aws-sdk-s3` gem.

eedwards-sk replied:
@amancevice No worries, actually your input is useful. I've been stumbling over this whole approach to requiring the S3 gem as it feels 'clunky' in the face of the lot of the magic that Homebrew does around isolating its environment and the bottle vs source install phases.

I've even thought maybe we could dump out the `GEM_PATH` and `GEM_HOME` values to help the user determine where that repo is.

Right now, there seem to be situations where the user needs to install the gem to the 'homebrew vendored ruby's gem repo', but there's no interface exposed by brew to do this, so instead they have to manually set their GEM_HOME / GEM_REPO paths.

eedwards-sk replied:
@amancevice this is a PR so it's not on the current homebrew, which means you are going to want the `aws-sdk-v1` gem,  not `aws-sdk-s3` (which is what I'm proposing to switch to with this PR).

For your case, try

`gem install aws-sdk-v1 --no-rdoc --no-ri --user-install`

and make sure it ends up in your user dir (`$HOME/.gem/`...)

amancevice replied:
I was actually writing my own `PresignDownloadStrategy` class so that I could use the newer gem instead of the `-v1` classic:

```ruby
class PresignDownloadStrategy < CurlDownloadStrategy
  def _fetch

    # Load AWS library
    begin
      require 'aws-sdk-s3'
    rescue LoadError
      # No idea what to do here...
      onoe "please install the aws-sdk-s3 gem into the gem repo used by brew"
    end

    # Get presigned URL
    signer = Aws::S3::Presigner.new
    uri = URI.parse @url
    url = signer.presigned_url :get_object,
                               :bucket => uri.host,
                               :key => uri.path.sub(/\A\//, '')

    # Download
    curl_download url, to: temporary_path
  end
end
```

I ended up junking it and writing my own code to compute a presigned S3 URL using the standard library (info on how to do that [here](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html)).

It works, but definitely not ideal. For one, anyone who wants to use it has to export the `HOMEBREW_AWS_ACCESS_KEY_ID` and `HOMEBREW_AWS_SECRET_ACCESS_KEY` environment variables, which having the SDK would avoid.

eedwards-sk replied:
Ah yeah, that's definitely not ideal because the big advantage (imo) of the S3 library is that it will do the credentials magic of looking in your ~/.aws folder, and supporting instance profile credentials present on the AWS instance metadata endpoint.

Having to export the actual keys means needing to do all that work themselves (e.g. CURLing the metadata endpoint for the instance credential keys and then setting the env vars).

eedwards-sk replied:
Also the presign strategy is basically exactly what the S3DownloadStrategy is already doing, although I like how you're using a presigner directly instead of relying on the object summary. I'll have to do some experimenting with that approach when I revisit this PR this weekend (I want to experiment with Mike's suggestion of installing gems in `formula_installer.rb`)

amancevice replied:
FWIW, this seemed to work for me:

```
GEM_HOME=$(brew --repo)/Library/Homebrew/vendor/portable-ruby/current/lib/ruby/gems/2.3.0 gem install aws-sdk-s3 --no-rdoc --no-ri
```

amancevice replied:
I ended up writing a fetch-only formula for my colleagues, as it seemed like the fetch environment is permissive-enough to allow the installation of gems into the Homebrew repo.

One coworker had a problem because somehow Homebrew was using his system Ruby so he had to `sudo gem install aws-sdk-s3` on his own, but everyone else got by with `brew fetch private/tap/aws-sdk`

```ruby
begin
  require "aws-sdk-s3"
  ohai "The aws-sdk-s3 is already installed at #{Gem::dir}"
rescue LoadError
  ohai "Installing the aws-sdk-s3 to #{Gem::dir}"
  ENV["GEM_HOME"] = ENV["GEM_PATH"] = Gem::dir
  begin
    Gem::install "aws-sdk-s3", "~> 1.8"
    require "aws-sdk-s3"
  rescue
    odie <<~EOS
      Something went wrong attempting to install the aws-sdk-s3 gem.

      Send the output of the following to someone in Engineering:

        brew config

    EOS
  end
end
```

MikeMcQuaid replied:
The solution for this is to use `install_gem_setup_path!` but outside of `download_strategy` e.g. `formula_installer.rb`. This will ensure the gem is always installed automatically and into the right location.

MikeMcQuaid replied:
I'd suggest that that be part of this PR but if you want to make a dedicated PR just for that @amancevice it would be much appreciated.

amancevice replied:
Thanks, @MikeMcQuaid -- I didn't mean to hijack the thread, just adding color to help @eedwards-sk & others decide how to handle this stuff.

MikeMcQuaid commented:
@eedwards-sk Any news on this?

eedwards-sk commented:
@MikeMcQuaid thanks for following up, I've been busy at work and let this slip :)

I'll jump on it tonight. I think I'm going to skip trying to install the gem automagically and leave that for a follow-up PR (e.g. maybe @amancevice wants to take a stab at it?)

MikeMcQuaid commented:
No worries and thanks both @eedwards-sk and @amancevice!

kyawlaykoko commented:
Myanmar

eedwards-sk commented:
@MikeMcQuaid 

Added latest changes with your suggestions. Note that I'm getting a failing test to something unrelated when I run `brew tests --online`, so I guess we'll see if it fails in travis.

As an aside, if I can get on my soapbox for a moment: this entire approach for testing and utilizing the S3 library feels like a total smell.

The tests I added, while adding some coverage, are basically just testing the implementation, and with the amount of mocking needed to not actually exercise the s3 library, I'm basically mocking the implementation, too.

I think a more holistic downstream improvement would include:

- moving the s3 utilization into a 'service' class
- use dependency injection to provide the service
- mock the service during tests
- write tests for the service, as well

The last line (tests for the service) would actually require a dependency on the s3 gem. I honestly **don't see how you can expect to utilize a 3rd party library and then not include it when testing it**. Without that, you're not actually exercising their contract AT ALL, so the tests are going to hide any changes in their API, which feels like a smell.

I'd like to see an 'officially blessed' approach to using 3rd party libraries or gems, maybe to make this an officially supported concept where we can treat bringing in gems to the 'brew repo' as a first-class concept. As of now it feels more like a hack if we proceed with trying to auto-install them.

Anyway, that's enough of my soap box. I hope this should at least take what already existed, improve it, and leave it easier to improve next time.

eedwards-sk commented:
The check failure appears to be completely unrelated to this PR. Is master currently broken?

amancevice commented:
Adding to @eedwards-sk’s soapbox: it may have made more sense to avoid installing the `aws-sdk-v1` gem because it pulled in some heavy requirements (hello, nokogiri), but the newer `aws-sdk-s3` gem is much slimmer. It only requires a handful of lightweight gems, mostly other aws stuff.

MikeMcQuaid commented:
> I think a more holistic downstream improvement would include

These seem overkill to me, I'm afraid. Additional indirect and abstraction makes the code harder to read and in my general experience integration tests catch far more than unit tests do.

> The last line (tests for the service) would actually require a dependency on the s3 gem. I honestly don't see how you can expect to utilize a 3rd party library and then not include it when testing it. Without that, you're not actually exercising their contract AT ALL, so the tests are going to hide any changes in their API, which feels like a smell.

I agree. That said I'm not sure how you can test a dependency without actually accessing S3. I'd be 👍 on something that made `brew tests --online` actually download something small from a public S3 bucket (e.g. one of Amazon's).

> I'd like to see an 'officially blessed' approach to using 3rd party libraries or gems, maybe to make bringing in gems to the 'brew repo' a first-class concept. As of now it feels more like a hack if we proceed with trying to auto-install them.

This won't happen I'm afraid. The only reason we're allowing it on this download strategy is because it's a pretty niche use-case. It's a feature rather than a bug that Homebrew works without requiring any non-vendored gems.

> but the newer aws-sdk-s3 gem is much slimmer. It only requires a handful of lightweight gems, mostly other aws stuff.

If it's really slim we can consider vendoring it in `Library/Homebrew/vendor/`.

---

Aside from the above this PR and the others are all looking good and thanks for all the work, folks. There's merge conflicts here that need resolved but they should be the last ones.

eedwards-sk commented:
Bah! There were no merge conflicts you just waited too long to merge :P

I'll fix this today.

amancevice commented:
Sorry, @eedwards-sk this is because of my contribs.

FYI, You shouldn't need to require the `aws-sdk-s3` gem in the `S3DownloadStrategy` anymore.

Also, feel free to overwrite my changes to continue using the `Aws::S3::ObjectSummary` method if you have reasons, but I think the `Aws::S3::Presigner` method is simpler.

Lastly, you may want to move the "Bad S3 URL" error up in the stack so it doesn't need to be run twice (once for the bucket parse & once for the key parse). This would also make it easier to add `s3://`-style URLs to the mix.

eedwards-sk commented:
@amancevice Would have been much easier if you had waited for this improvement before getting yours merged.

I can't commit to making those changes soon, I may need to revisit this later. Maybe you should take over the PR.

eedwards-sk commented:
@MikeMcQuaid This was ready to be merged with no conflicts prior to you merging their PR which then blocked it...

As I said above, I expected the gem installation improvement to be a follow-up PR:

>and leave that for a follow-up PR (e.g. maybe @amancevice wants to take a stab at it?)

Thus my point. Appreciate the improvements either way, but I believe my point is valid. That's all I'll say on the topic. Thanks.

MikeMcQuaid commented:
> This was ready to be merged with no conflicts prior to you merging their PR which then blocked it...

I did not consider it ready to be merged yet or I would have merged it, sorry.

amancevice commented:
@eedwards-sk the following should resolve the conflicts:

Rebase from master:

```bash
git rebase master -s recursive -X theirs
```

Download [GH-3838.txt](https://github.com/Homebrew/brew/files/1843293/GH-3838.txt) to `~/GH-3838.txt`

Apply the diff:

```bash
git apply ~/GH-3838.txt
```

Review & commit the diff, then force-push onto your feature branch. That should resolve the conflicts. It's up to you if you want to make the changes I recommended above.

eedwards-sk commented:
@amancevice The scope of this PR was originally quite simple but was ballooned excessively over time due to asks for 'nice to haves', so at this point I am not interested in iterating on it any further until we get the core improvement (bottle support) merged.

If this work is not mergeable as-is then I wholly support anyone iterating on it further as necessary, but I won't have time to address it again for at least 2 weeks.

MikeMcQuaid commented:
> The scope of this PR was originally quite simple but was ballooned excessively over time due to asks for 'nice to haves', so at this point I am not interested in iterating on it any further until we get the core improvement (bottle support) merged.

@eedwards-sk I've tried to show some restrain so far as you're a first-time contributor and we appreciate your contribution but to be clear: this PR was not merged because it was not ready and all the things you consider "nice to haves" I consider requirements to merge this functionality (which is why it was not merged). You appear to be relatively new to GitHub open source contribution and I hope you continue to do more but arguing with maintainers about how they run their project and criticising other contributors on your first PR is not a recipe for a productive interaction.

MikeMcQuaid commented:
> If this work is not mergeable as-is

It's not. If you look at the diff it's pretty obviously now including changes that are currently in the `master` branch.

eedwards-sk commented:
@MikeMcQuaid 

>this PR was not merged because it was not ready and all the things you consider "nice to haves" I consider requirements to merge this functionality (which is why it was not merged)

The feature I've touched here was not my creation -- this is already part of homebrew, and is missing tests and is basically half baked (working only on source URLs and not bottles).

If having it 'auto install the gem' and adding more coverage were required functionality then the initial PR(s) that added S3DownloadStrategy should have been addressed as such.

It's also no longer clear what the actual requirements are to get this merged, you've simply said twice you would not merge it without explaining specifically.

Please keep that context in mind when considering why I might be frustrated.

MikeMcQuaid commented:
> basically half baked (working only on source URLs and not bottles).

That's not half-baked, that's something that applies to literally every download strategy.

> If having it 'auto install the gem' and adding more coverage were required functionality then the initial PR(s) that added S3DownloadStrategy should have been addressed as such.

This is not how open source works. Your changes revealed a deficiency (in code which you did have to change) which needed addressed before this was ready to merge.

> It's also no longer clear what the actual requirements are to get this merged, you've simply said twice you would not merge it without explaining specifically.

That's because you've said how busy you are and are unwilling to make further changes for several weeks for this PR so I didn't see the point in doing so until the existing issues (merge conflicts) had been addressed.

Again, advice on contributing to open source: the maintainers are need to review and merge (and create) more PRs than you are contributing so please try not to waste their time or your time arguing with them and telling them they are wrong.
